[DOC/TST] New docstring testing package `darglint`

loganthomas opened the pull request:
<!-- Thank you for your PR!

BEFORE YOU CONTINUE! Please add the appropriate three-letter abbreviation to your title.

The abbreviations can be:
- [DOC]: Documentation fixes.
- [ENH]: Code contributions and new features.
- [TST]: Test-related contributions.
- [INF]: Infrastructure-related contributions.

Also, do not forget to tag the relevant issue here as well.

Finally, as commits come in, don't forget to regularly rebase!
-->

# PR Description

Please describe the changes proposed in the pull request:

- This PR implements the [`darglint`](https://github.com/terrencepreilly/darglint) package: a functional docstring linter which checks whether a docstring's description matches the actual function/method implementation.

<!-- Doing so provides maintainers with context on what the PR is, and can help us more effectively review your PR. -->

<!-- Please also identify below which issue that has been raised that you are going to close. -->

**This PR resolves #745 .**

<!-- As you go down the PR template, please feel free to delete sections that are irrelevant. -->

# PR Checklist

<!-- This checklist exists for newcomers who are not yet familiar with our requirements. If you are experienced with
the project, please feel free to delete this section. -->

Please ensure that you have done the following:

1. [X] PR in from a fork off your branch. Do not PR from `<your_username>`:`dev`, but rather from `<your_username>`:`<feature-branch_name>`.
<!-- Doing this helps us keep the commit history much cleaner than it would otherwise be. -->
2. [X] If you're not on the contributors list, add yourself to `AUTHORS.rst`.
<!-- We'd like to acknowledge your contributions! -->
3. [X] Add a line to `CHANGELOG.rst` under the latest version header (i.e. the one that is "on deck") describing the contribution.
    - Do use some discretion here; if there are multiple PRs that are related, keep them in a single line.

## Quick Check

To do a very quick check that everything is correct, follow these steps below:

- [X] Run the command `make check` from pyjanitor's top-level directory. This will automatically run:
    - black formatting
    - flake8 checking
    - running the test suite
    - docs build

Once done, please check off the check-box above.

If `make check` does not work for you, you can execute the commands listed in the Makefile individually.

## Code Changes

None

## Documentation Changes

<!-- If you have not made documentation changes, please feel free to delete this section. -->

If you are adding documentation changes, please ensure the following:

- [X] Build the docs locally.
- [X] View the docs to check that it renders correctly.

# Relevant Reviewers

<!-- Finally, please tag relevant maintainers to review. -->

Please tag maintainers to review.

- @ericmjl

loganthomas committed:
[INF] initial commit of .darglint file

loganthomas committed:
[INF] add darglint to requirements-dev.txt and pre-commit hooks

loganthomas committed:
[INF] update darglint pre-commit hook to use verbosity of 2

loganthomas committed:
[DOC] update functions.move() docstring with typehint references and raise clauses

loganthomas committed:
[DOC] update tests/functions/test_also.drop_inplace() to ignore darglint checking

loganthomas committed:
[DOC] update scripts/check-autodoc.extract_folder_names() to be single line docstring for darglint

loganthomas committed:
[DOC] update janitor/spark/backend register_dataframe funcs to ignore darglint checking

loganthomas committed:
[DOC] update janitor/testing_utils/strategies.df_strategy() to ignore darglint checking

loganthomas committed:
[DOC] update math.ecdf() with raise clauses

loganthomas committed:
[DOC/FIX] update functions.move() to encapsulate raise clauses with raise type

loganthomas committed:
[DOC] update janitor/xarray/functions.clone_using() docstring with raise clauses

loganthomas committed:
[DOC] update janitor/math.py with darglint docstrings

* math.log() to ignore ``error`` as optional input

* math.sigmoid() to ignore multi-line docstring

* math.logit() to ignore ``error`` as optional input
* math.logit() to include ``raises`` clause

* math.probit() to ignore ``error`` as optional input

* math.z_score() to ignore ``moments_dict`` as optional input

loganthomas committed:
[DOC/FIX] add double space for noqa statements in docstrings

loganthomas committed:
[DOC] update engineering.convert_units with raise clause

loganthomas committed:
[DOC] update chemistry.morgan_fingerprint() with raise clause

loganthomas committed:
[DOC] update chemistry.smiles2mol() with raise clause

loganthomas committed:
[DOC] update utils.py with raise clauses

loganthomas committed:
[DOC] update setup.pt generate_long_description() with raise clause

loganthomas committed:
[DOC] update tests/xarray/contest.py da() docstring to ignore darglint

loganthomas committed:
[DOC] upates finance.py docstrings with return clauses

loganthomas committed:
Merge remote-tracking branch 'upstream/dev' into darglint_testing_#745

loganthomas committed:
[DOC] upate get_feature_targets() docstring in ml.py to not reference arg types

loganthomas committed:
[DOC] update read_csvs() docstring in io.py with return and raises clauses

loganthomas committed:
[DOC] update complete() docstring in functions.py to encapsulate raise clauses in colons

loganthomas committed:
[DOC] update groupby_topk() docstring in functions.py to encapsulate raise clauses in colons

loganthomas committed:
[DOC] update fill_direction() docstring in functions.py to encapsulate raise clauses in colons

loganthomas committed:
[DOC] update process_text() docstring in functions.py to encapsulate raise clauses in colons and include arg, kwarg, str_function params

loganthomas committed:
[DOC] test if noqa shows up in docs build

loganthomas committed:
[DOC] try to hide noqa with ..

loganthomas committed:
[DOC] try to hide noqa with .. and new line

loganthomas committed:
[DOC] upadte darglint ignores in docstiong with new line and preceeding .. so ignores dont show up in docs

loganthomas committed:
[DOC] update expand_grid() docstring in functions.py to encapsulate raise clauses in colons

loganthomas committed:
[DOC] update sort_naturally() docstring in functions.py with returns clause

loganthomas committed:
[DOC] update jitter() docstring in functions.py with raise clauses

loganthomas committed:
[DOC] update join_apply() docstring in functions.py with returns clause

loganthomas committed:
[DOC] update shuffle() docstring in functions.py with returns clause

loganthomas committed:
[DOC] update drop_duplicate_columns() docstring in functions.py with 4 space indent

loganthomas committed:
[DOC] update bin_numeric() docstring in functions.py with raises clause

loganthomas committed:
[DOC] update DataDescitption.set_description() docstring in functions.py with raises clause

loganthomas committed:
[DOC] update groupby_agg() docstring in functions.py by deletiing axis param

loganthomas committed:
[DOC] update find_replace() docstring in functions.py with returns clause

loganthomas committed:
[DOC] update also() docstring in functions.py with args and kwargs param clauses

loganthomas committed:
[DOC] update iimpute() docstring in functions.py with raise clauses

loganthomas committed:
[DOC] update select_columns() docstring in functions.py with raise clauses

loganthomas committed:
[DOC] update min_max_scale() docstring in functions.py and ValueError

loganthomas committed:
[DOC] update transform_columns() docstring in functions.py with raises clause

loganthomas committed:
[DOC] update add_column() docstring in functions.py with raise clauses and update ValueError messages

loganthomas committed:
[DOC] update change_type() docstring in functions.py with raise clause

loganthomas committed:
[DOC] update filter_column_isin() docstring in functions.py with raises clause

loganthomas committed:
[DOC] update deconcatenate_column() docstring in functions.py with raises clause

loganthomas committed:
[DOC] update concatenate_columns() docstring in functions.py with raise clause

loganthomas committed:
[DOC] update expand_column() docstring in functions.py to leave out type defn for concat param

loganthomas committed:
[DOC] update fill_empty() docstring in functions.py with raises clause

loganthomas committed:
[DOC] update convert_unix_date() docstring in functions.py to leave out type defn for column_name param

loganthomas committed:
[DOC] update convert_matlab_date() docstring in functions.py to leave out type defn for column_name param

loganthomas committed:
[DOC] update convert_excel_date() docstring in functions.py to leave out type defn for column_name param

loganthomas committed:
[DOC] update convert_excel_date() docstring in functions.py to leave out type defn for params

loganthomas committed:
[DOC] update reorder_columns() docstring in functions.py with raises clause

loganthomas committed:
[FIX] change typo in reorder_columns() raises clause

loganthomas committed:
[DOC] update label_encode() docstring in functions.py with raise clauses

loganthomas committed:
[DOC] update encode_categorical() docstring in functions.py with raise clauses

loganthomas committed:
[DOC] update get_dupes() docstring in functions.py to leave out type defn for column_name param

loganthomas committed:
[DOC] update clean_names() docstring in functions.py with strip_accents param

loganthomas committed:
[DOC] update unionize_dataframe_categories() in functions.py with raises clause

loganthomas committed:
[INF] remove .darglint file and move configs to pyproject.toml

loganthomas committed:
[INF] delete .darglint file

loganthomas committed:
[INF/FIX] wrap darglint configs in quotes

loganthomas committed:
[INF/FIX] fix quote typo

loganthomas committed:
[INF] revert back to .dargline file as docs say must exist in repo

loganthomas committed:
[FIX] .darglint should have not str values

loganthomas committed:
Merge remote-tracking branch 'upstream/dev' into darglint_testing_#745

loganthomas committed:
[DOC] add param and return to functions.patterns()

loganthomas committed:
[DOC] add line to changelog

loganthomas commented on .darglint:
```diff
@@ -0,0 +1,4 @@
+[darglint]
+docstring_style=sphinx
+strictness=short
+ignore_regex=^(test_|_)(.*)
```
``darglint`` requires its own settings file. I tried implementing this in the `pyproject.tml` but it didn't work.

From the ``darglint`` docs:
> ``darglint`` can be configured using a configuration file. The configuration file must be named either .darglint, setup.cfg, or tox.ini. It must also have a section starting with the section header, [darglint]. Finally, the configuration file must be located either in the directory darglint is called from, or from a parent directory of that working directory.

ericmjl replied:
Got it! I chimed in [on the `darglint` repo](https://github.com/terrencepreilly/darglint/issues/130) to ask if pyproject.toml is on the roadmap. Meanwhile, feel free to resolve this thread. :smile:.

loganthomas commented on .pre-commit-config.yaml:
```diff
@@ -21,5 +21,10 @@ repos:
+  rev: master
+  hooks:
+  - id: darglint
+    args: [-v 2]
```
``args: [-v 2]`` makes the error messages a bit less cryptic.

Without ``-v 2``:
```bash
functions.py:unionize_dataframe_categories:49: DAR401: -r TypeError
finance.py:convert_currency:409: DAR201: - return
```

With ``-v 2``:

```bash
functions.py:unionize_dataframe_categories:49: DAR401: Missing exception(s) in Raises section: -r TypeError
finance.py:convert_currency:409: DAR201: Missing "Returns" in Docstring: - return
```

ericmjl replied:
Cool stuff! Could I suggest adding a comment here too?

```suggestion
    args: [-v 2]  # this config makes the error messages a bit less cryptic.
```

loganthomas commented on requirements-dev.txt:
```diff
@@ -19,3 +19,4 @@ unyt
 xarray
 pandas-vet
 interrogate
+darglint
```
@ericmjl 

I added this to the ``requirements-dev.txt`` file but wasn't sure if this should also be placed in the `requirements.txt` file as well 🤷‍♂️ .

I imagine we might want to include this is a test pipeline (much like we are doing with ``black`` and ``interrogate``).

ericmjl replied:
> I added this to the requirements-dev.txt file but wasn't sure if this should also be placed in the requirements.txt file as well 🤷‍♂️ .

I think that might not be necessary, unless something in the CI system fails without `darglint` in the `requirements.txt` file.

> I imagine we might want to include this is a test pipeline (much like we are doing with black and interrogate).

In the testing pipelines, the development environment should be specified fully using `environment-dev.yml`. That is the configuration file that is used in Azure pipelines.

loganthomas replied:
Ok great. Should I add something to the pipeline then? Do we want to have `darglint` run?

```
# Add to .azure-pipelines/templates/run-tests.yml??
  - script: |
      $(activate.command) pyjanitor-dev
      darglint . -v 2
```

ericmjl replied:
Yep, let's do that! It'll help as a backup to the pre-commit hooks. :smile:.

loganthomas replied:
ok, great. Just pushed and its running now... I have it set up to run only on the `janitor` dir. Did you want it to run on the root dir? I figure the only thing we are really concerned with as far as docstring linting is the core package.

```
# Current config
- script: |
      $(activate.command) pyjanitor-dev
      darglint janitor -v 2
```

```
# Update if you'd like...
- script: |
      $(activate.command) pyjanitor-dev
      darglint . -v 2
```

loganthomas replied:
Resolving as we discussed this on the _realease party_ call for v0.20.10. We decided to run `darglint` on janitor (not the root)... so no changes needed.

loganthomas commented on janitor/functions.py:
```diff
@@ -4154,10 +4213,14 @@ def process_text(
+    :raises KeyError: if ``string_function`` is not a Pandas string method.
+    :raises TypeError: if wrong ``arg`` or ``kwarg`` is supplied.
+
+    .. # noqa: DAR402
```
This allows ``darglint`` to leave out checking for specific errors. The ``..`` prefix ensures that the ``noqa: DAR402`` doesn't appear in the docs.

Here, a DAR402 is raised (_DAR402_: The docstring describes an exception not explicitly raised.) since the function _could_ raise a TypeError, but doesn't explicitly have one in the source code. We use this to ignore the ``darglint`` raised error.

loganthomas commented on janitor/functions.py:
```diff
@@ -4045,8 +4104,8 @@ def expand_grid(
     :returns: A pandas dataframe of all combinations of name value pairs.
-    :raises: TypeError if others is not a dictionary
-    :raises: KeyError if there is a dataframe and no key is provided.
+    :raises TypeError: if others is not a dictionary
+    :raises KeyError: if there is a dataframe and no key is provided.
```
Just an exmaple that ``raises`` clauses should include the type of error raised between colons. This is pretty cool because it automatically puts a link to the Pythons type errors in the docs.

After making this change, the docs show links to [TypeError](https://docs.python.org/3/library/exceptions.html#TypeError) and [KeyError](https://docs.python.org/3/library/exceptions.html#KeyError)


![Screen Shot 2020-10-01 at 08 46 29](https://user-images.githubusercontent.com/24984410/94817432-9d263e00-03c2-11eb-9955-bd25d68fcc69.png)

samukweku replied:
I think I like this one. Helpful for users. How do you create the hyperlink?

loganthomas replied:
It's built in and automatic 😄 As long as the ``raises`` clause includes the error *within* the colons.

samukweku replied:
@loganthomas , this is really cool. Definitely have to play with this in my next PR. If I get stuck, I'll reach out to you.

ericmjl reviewed (changes requested):
Just one suggested change!

loganthomas committed:
[DOC] add comment to pre-commit configs explaining darglint args

loganthomas committed:
[FIX] line 1118 in functions.py causes pipeline fail; line too long

loganthomas committed:
[INF] add darglint to run-tests.yml

loganthomas committed:
[INF/FIX] fix typo in run-tests.yml for darglint step...doh

loganthomas committed:
[INF] sort requirements-dev.txt

loganthomas committed:
[INF] add darglint; remove duplicate pandas-vet from requirements.txt

loganthomas commented:
@ericmjl 
- Added the comment you suggested `.pre-commit-config.yaml`
- Updated `.azure-pipelines/templates/run-tests.yml` to run a darglint stage
- Added `darglint` to __both__ `requirements-dev.txt` and `requirements.txt` (pipelines failed and couldn't "find darglint" without it in the `requirements.txt` file.)

The only hanging comment is whether you want `darglint` to run at the root of `pyjanitor` or pointed directly at `janitor`. This was a fun one. I liked testing this out 🤓

loganthomas committed:
Merge remote-tracking branch 'upstream/dev' into darglint_testing_#745

loganthomas committed:
[DOC] update changelog after new release

ericmjl commented:
I approve! Thanks @loganthomas :smile:
